### ISSUE TITLE
`needless_pass_by_value`: Wrongly emit on Rc values

### ISSUE THREAD

**Issue by dev-ardi** [rust-lang/rust-clippy#13155]
### Summary

It's more expensive to pass in a `&Rc` than just cloning it.

### Lint Name

clippy::needless_pass_by_value

### Reproducer

I tried this code:

```rust
#![warn(clippy::needless_pass_by_value)]
#![allow(unused)]
use std::rc::Rc;

fn foo(s: Rc<()>) {}
```

I saw this happen:

```
warning: this argument is passed by value, but not consumed in the function body
 --> src/main.rs:5:11
  |
5 | fn foo(s: Rc<()>) {}
  |           ^^^^^^ help: consider taking a reference instead: `&Rc<()>`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_value
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::needless_pass_by_value)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

### Version

```text
latest
```

**Comment by lukaslueg** [comment]
> It's more expensive to pass in a &Rc than just cloning it.

How do we arrive at this general conclusion? "Expensive" in what way?

To the contrary: Passing a `&Rc<T>` is completely free, and it's hard to beat that. On the other hand, passing a clone requires at the very least one comparison and a branch (to check if the refcount overflows), and almost certainly adds code to every scope where that value gets dropped in order to run the `Rc`'s destructor (adding more comparisons, more branches, more instructions).

**Comment by JP-Ellis** [comment]
I do wonder whether `needless_pass_by_value` should have the ability to be toggled based on the type, especially to handle cases of smart pointers.

I think it might be good to add the ability to configure the applicability of `needless_pass_by_value` based on the type. For example:

```toml
allow-pass-by-value = ["std::rc::Rc", "std::sync::Arc", "crate::MySmartPointer"]
```

**Comment by Jarcho** [comment]
In general taking `&Rc<T>` as an argument will result in a single instruction to read the contained `RcInner` pointer. For x86 this would be something like `mov rax, [rax]`. By taking `Rc<T>` there will at least be at least one `drop` call which is more expensive than the previous pointer read, plus you have a `clone` call at each callsite. I'm honestly not sure what circumstances would lead to `Rc<T>` being faster while also triggering `needless_pass_by_value`.

Now is taking `Rc<T>` instead of `&Rc<T>` meaningfully slower? Almost certainly not when compared to everything else you're doing with the `Rc`.

**Comment by JP-Ellis** [comment]
In that case, should this issue be closed?
